### ISSUE TITLE
[ONNX] Systematically expand ONNX functions before conversion to Torch to avoid needing bespoke conversions

### ISSUE THREAD

**Issue by andfau-amd** [llvm/torch-mlir#3384]
Currently there are [a large number of operators](https://github.com/nod-ai/SHARK-Turbine/issues/215) in ONNX that aren't supported by TorchOnnxToTorch. As a beginner, I was tasked with adding a conversion for one of those, namely [MeanVarianceNormalization](https://onnx.ai/onnx/operators/onnx__MeanVarianceNormalization.html), in https://github.com/nod-ai/SHARK-Turbine/issues/697. But I noticed that this operator is actually a function, and operators that are functions are defined by [a series of other ONNX operators](https://onnx.ai/onnx/intro/concepts.html#functions). In this case, all those other operators are ones we already have conversions for. So I wondered why we don't expand these operator functions before converting, to avoid needing bespoke conversions for all of them.

I discussed a bit with some colleagues (@ScottTodd, @rsuderman) and it doesn't seem like there's a good reason not to do this, so I'm going to make an attempt at it. Here's some things we've learned that might be relevant:

- The ONNX API has a function called [`FunctionExpandHelper`](https://github.com/onnx/onnx/blob/093a8d335a66ea136eb1f16b3a1ce6237ee353ab/onnx/defs/function.cc#L18) ~~(in Python:  `function_expand_helper`)~~ that seems to be intended for exactly this kind of expansion. Here are some places it has been used:
  - ONNX [uses _a similar Python function_ for generating expanded versions of tests](https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/backend/test/case/node/__init__.py#L154-L164). For the example of MeanVarianceNormalization you can follow the flows of:
    -  [original Python](https://github.com/onnx/onnx/blob/main/onnx/backend/test/case/node/meanvariancenormalization.py) -> (no expansion) -> [generated ONNX](https://github.com/onnx/onnx/blob/main/onnx/backend/test/data/node/test_mvn/model.onnx) -> [MLIR](https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/onnx/node/generated/test_mvn/model.mlir)
    -  [original Python](https://github.com/onnx/onnx/blob/main/onnx/backend/test/case/node/meanvariancenormalization.py) -> (expansion) -> [generated ONNX](https://github.com/onnx/onnx/blob/main/onnx/backend/test/data/node/test_mvn_expanded/model.onnx) -> [MLIR](https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/onnx/node/generated/test_mvn_expanded/model.mlir)
  -  ONNX also has a [basic test of it](https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/test/cpp/function_verify_test.cc#L342)
  - OpenVINO [uses it in its ONNX frontend](https://github.com/openvinotoolkit/openvino/blob/00856cdcb21de2a966b8e48cf64cf26041cbe06c/src/frontends/onnx/frontend/src/core/transform.cpp#L60)
-  ONNX also has an [inlining API](https://github.com/onnx/onnx/blob/093a8d335a66ea136eb1f16b3a1ce6237ee353ab/onnx/inliner/inliner.h#L41-L51) but it's not clear to me if it's suitable for this use.
- Trivia: MeanVarianceNormalization seems to be the poster child for an ONNX operator that is a function! The [guide on adding new ops](https://onnx.ai/onnx/repo-docs/AddNewOp.html) gives it as an example of something that should be a function rather than a primitive operator; it predates functions actually being operators and then got [promoted](https://github.com/onnx/onnx/commit/3a9a87871021a02d6b970c1e26ffa4c0dd720677#diff-edac4191612f111be4f5d6f83d690f8d15236c8068161c281f6ebe70cec81d93R8894); and it is even used for [testing function expansion](https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/test/cpp/function_verify_test.cc#L342). :)

Regarding actually implementing an expansion in our importer or converter, some things to keep track of:

- Once we have expansion working, there should be an assertion somewhere to catch cases where operators we expect to be expanded are somehow still present in the IR.
- In case we want to have a bespoke conversion for some particular operator, there should be some allowlist/denylist setup to prevent some operators being expanded, so they can be converted later.
- There should be some instrumentation on the expansion, so it's easy to see which operators are getting expanded, enabling decision-making about possible future special-casing.

If we do implement an expansion, then:

- Should we get rid of any existing bespoke lowerings that are now redundant? I know there's at least one. Maybe something for follow-up work.

**Comment by andfau-amd** [comment]
> The ONNX API has a function called FunctionExpandHelper **(in Python: `function_expand_helper`)**

> ONNX **uses [the Python version of] it for generating expanded versions of tests**

Oops, that's not the same function! I'd missed this:

https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/backend/test/case/node/__init__.py#L92-L95

So this Python variant of this function only exists internally to this test code. This might not be good for the idea of doing the expansion in our Python converter (`python/torch_mlir/tools/import_onnx/__main__,py`), but we'll see.

**Comment by stellaraccident** [comment]
I haven't looked at this super closely, but in my mind at the beginning, I thought the importer could just be taught to import such functions as private func.func and then emit a func.call to them when used. I think that would match semantics and leave most of the work in MLIR.

**Comment by stellaraccident** [comment]
Yeah, this is how some other frontends do it: get it in mlir and let the infra there take over. 

Also keeps the importer mechanical.

**Comment by andfau-amd** [comment]
It was more complicated than I expected, and I might have missed some things, but I managed to get something working!

I can get it to spit out an MLIR function for the relevant ONNX operator and call it:
```
python -m torch_mlir.tools.import_onnx mlir_venv/lib/python3.10/site-packages/onnx/backend/test/data/node/test_mvn/model.onnx
```
```mlir
module {
  func.func @test_mvn(%arg0: !torch.vtensor<[3,3,3,1],f32>) -> !torch.vtensor<[3,3,3,1],f32> attributes {torch.onnx_meta.ir_version = 7 : si64, torch.onnx_meta.opset_version = 13 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
    %none = torch.constant.none
    %0 = call @"('MeanVarianceNormalization', '', 18, [tensor_type {\0A  elem_type: 1\0A  shape {\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 1\0A    }\0A  }\0A}\0A], [tensor_type {\0A  elem_type: 1\0A  shape {\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 1\0A    }\0A  }\0A}\0A], [])"(%arg0) : (!torch.vtensor<[3,3,3,1],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    return %0 : !torch.vtensor<[3,3,3,1],f32>
  }
  func.func @"('MeanVarianceNormalization', '', 18, [tensor_type {\0A  elem_type: 1\0A  shape {\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 1\0A    }\0A  }\0A}\0A], [tensor_type {\0A  elem_type: 1\0A  shape {\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 3\0A    }\0A    dim {\0A      dim_value: 1\0A    }\0A  }\0A}\0A], [])"(%arg0: !torch.vtensor<[3,3,3,1],f32>) -> !torch.vtensor<[3,3,3,1],f32> attributes {torch.onnx_meta.ir_version = 0 : si64, torch.onnx_meta.opset_version = 18 : si64, torch.onnx_meta.producer_name = "", torch.onnx_meta.producer_version = ""} {
    %none = torch.constant.none
    %0 = torch.operator "onnx.Constant"() {torch.onnx.value = dense<2.000000e+00> : tensor<f32>} : () -> !torch.vtensor<[],f32>
    %1 = torch.operator "onnx.Constant"() {torch.onnx.value = dense<9.99999971E-10> : tensor<f32>} : () -> !torch.vtensor<[],f32>
    %2 = torch.operator "onnx.Constant"() {torch.onnx.value_ints = [0 : si64, 2 : si64, 3 : si64]} : () -> !torch.vtensor<[3],si64>
    %3 = torch.operator "onnx.ReduceMean"(%arg0, %2) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[3],si64>) -> !torch.vtensor<[1,3,1,1],f32>
    %4 = torch.operator "onnx.Pow"(%3, %0) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %5 = torch.operator "onnx.Pow"(%arg0, %0) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    %6 = torch.operator "onnx.ReduceMean"(%5, %2) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[3],si64>) -> !torch.vtensor<[1,3,1,1],f32>
    %7 = torch.operator "onnx.Sub"(%6, %4) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %8 = torch.operator "onnx.Sqrt"(%7) : (!torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %9 = torch.operator "onnx.Sub"(%arg0, %3) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    %10 = torch.operator "onnx.Add"(%8, %1) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %11 = torch.operator "onnx.Div"(%9, %10) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    return %11 : !torch.vtensor<[3,3,3,1],f32>
  }
}
```

I apologise for the very ugly function name. Hopefully a slightly more pleasant name mangling scheme is possible. :)

Then if I change the importer so that the main function is marked as public, and the operator's function is marked as private, MLIR can inline it and throw it away:

```
python -m torch_mlir.tools.import_onnx mlir_venv/lib/python3.10/site-packages/onnx/backend/test/data/node/test_mvn/model.onnx | build/bin/torch-mlir-opt --split-input-file --inline
```
```mlir
module {
  func.func public @test_mvn(%arg0: !torch.vtensor<[3,3,3,1],f32>) -> !torch.vtensor<[3,3,3,1],f32> attributes {torch.onnx_meta.ir_version = 7 : si64, torch.onnx_meta.opset_version = 13 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
    %0 = torch.operator "onnx.Constant"() {torch.onnx.value = dense<2.000000e+00> : tensor<f32>} : () -> !torch.vtensor<[],f32>
    %1 = torch.operator "onnx.Constant"() {torch.onnx.value = dense<9.99999971E-10> : tensor<f32>} : () -> !torch.vtensor<[],f32>
    %2 = torch.operator "onnx.Constant"() {torch.onnx.value_ints = [0 : si64, 2 : si64, 3 : si64]} : () -> !torch.vtensor<[3],si64>
    %3 = torch.operator "onnx.ReduceMean"(%arg0, %2) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[3],si64>) -> !torch.vtensor<[1,3,1,1],f32>
    %4 = torch.operator "onnx.Pow"(%3, %0) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %5 = torch.operator "onnx.Pow"(%arg0, %0) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    %6 = torch.operator "onnx.ReduceMean"(%5, %2) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[3],si64>) -> !torch.vtensor<[1,3,1,1],f32>
    %7 = torch.operator "onnx.Sub"(%6, %4) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %8 = torch.operator "onnx.Sqrt"(%7) : (!torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %9 = torch.operator "onnx.Sub"(%arg0, %3) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    %10 = torch.operator "onnx.Add"(%8, %1) : (!torch.vtensor<[1,3,1,1],f32>, !torch.vtensor<[],f32>) -> !torch.vtensor<[1,3,1,1],f32>
    %11 = torch.operator "onnx.Div"(%9, %10) : (!torch.vtensor<[3,3,3,1],f32>, !torch.vtensor<[1,3,1,1],f32>) -> !torch.vtensor<[3,3,3,1],f32>
    return %11 : !torch.vtensor<[3,3,3,1],f32>
  }
}
```

And this looks fairly similar at first glance to the [known-good expansion](https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/onnx/node/generated/test_mvn_expanded/model.mlir), though I'm not yet sure if it's perfectly equivalent.

You can see what I have so far at https://github.com/llvm/torch-mlir/compare/main...andfau-amd:torch-mlir:onnx-to-torch-function-expansion. The interesting stuff is in `get_operator_function`, the rest is mostly plumbing.

(Perhaps I should make a draft PR? I think it would be useful to get some early feedback.)

**Comment by andfau-amd** [comment]
Resolved my FIXMEs, now it perfectly matches for MeanVarianceNormalization:

```
$ python -m torch_mlir.tools.import_onnx mlir_venv/lib/python3.10/site-packages/onnx/backend/test/data/node/test_mvn/model.onnx | build/bin/torch-mlir-opt --split-input-file --inline --convert-torch-onnx-to-torch --cse --canonicalize > mvn-ours.txt
$ python -m torch_mlir.tools.import_onnx mlir_venv/lib/python3.10/site-packages/onnx/backend/test/data/node/test_mvn_expanded/model.onnx | build/bin/torch-mlir-opt --split-input-file --inline --convert-torch-onnx-to-torch --cse --canonicalize > mvn-theirs.txt
$ sed -e 's/_expanded//g' mvn-theirs.txt > mvn-theirs2.txt
$ diff mvn-ours.txt mvn-theirs2.txt
$
```

I'll try to add some tests and then I'll open a pull request.

**Comment by Groverkss** [comment]
It's probably fine to not create a new function that is intended to be inlined anyway. You can replace it in place.

**Comment by renxida** [comment]
This is awesome

**Comment by andfau-amd** [comment]
Thank you!

**Comment by andfau-amd** [comment]
I've now gotten to the point where `cmake --build build --target check-torch-mlir` no longer fails with my patch. Lots of edge cases had to be fixed that were causing the importer to crash in pre-existing regression tests. I think I know too much about ONNX now. :) I have actually had to [denylist two operators](https://github.com/andfau-amd/torch-mlir/blob/2ee6c7aa9a24356e6f827293d2c419a7f91337d7/python/torch_mlir/extras/onnx_importer.py#L742-L759) to prevent them being expanded, for fairly good reasons. :(

With that out of the way I can actually start adding new tests.

**Comment by andfau-amd** [comment]
Created a PR to track things beyond this point and allow commenting on the code etc. It's not quite ready for merging yet though. https://github.com/llvm/torch-mlir/pull/3409

**Comment by andfau-amd** [comment]
Marked the PR as ready for review. It still needs some more testing but that can be done in parallel I hope.

**Comment by andfau-amd** [comment]
Follow-up issue for expanding the allowlist: https://github.com/llvm/torch-mlir/issues/3464
